controllers/controller/workspacerouting/solvers/common.go

```diff
@@ -68,50 +75,69 @@ func getDiscoverableServicesForEndpoints(endpoints map[string]controllerv1alpha1
 	return services
 }
 
-func getServicesForEndpoints(endpoints map[string]controllerv1alpha1.EndpointList, meta WorkspaceMetadata) []corev1.Service {
-	if len(endpoints) == 0 {
-		return nil
+// GetServiceForEndpoints returns a single service that exposes all endpoints of given exposure types, possibly also including the discoverable types.
+func GetServiceForEndpoints(endpoints map[string]controllerv1alpha1.EndpointList, meta WorkspaceMetadata, includeDiscoverable bool, exposureType ...devworkspace.EndpointExposure) corev1.Service {
+	service := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      common.ServiceName(meta.WorkspaceId),
+			Namespace: meta.Namespace,
+			Labels: map[string]string{
+				config.WorkspaceIDLabel: meta.WorkspaceId,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: meta.PodSelector,
+			Type:     corev1.ServiceTypeClusterIP,
+		},
 	}
-	var servicePorts []corev1.ServicePort
-	for _, machineEndpoints := range endpoints {
-		for _, endpoint := range machineEndpoints {
-			isExposed := func(port int) bool {
-				for _, port := range servicePorts {
-					if port.TargetPort.IntVal == int32(endpoint.TargetPort) {
-						//port is already exposed
-						return true
-					}
-				}
-				return false
+
+	// "set" of ports that are still left for exposure
+	ports := map[int]bool{}
+	for _, es := range endpoints {
+		for _, endpoint := range es {
+			ports[endpoint.TargetPort] = true
+		}
+	}
+
+	// "set" of exposure types that are allowed
+	validExposures := map[v1alpha2.EndpointExposure]bool{}
+	for _, exp := range exposureType {
+		validExposures[exp] = true
+	}
+
+	for _, es := range endpoints {
+		for _, endpoint := range es {
+			if !validExposures[endpoint.Exposure] {
+				continue
 			}
 
-			if !isExposed(endpoint.TargetPort) {
-				servicePort := corev1.ServicePort{
+			if !includeDiscoverable && endpoint.Attributes.GetBoolean(string(controllerv1alpha1.DISCOVERABLE_ATTRIBUTE), nil) {
+				continue
+			}
+
+			if ports[endpoint.TargetPort] {
+				// make sure we don't mention the same port twice
+				ports[endpoint.TargetPort] = false
+				service.Spec.Ports = append(service.Spec.Ports, corev1.ServicePort{
 					Name:       common.EndpointName(endpoint.Name),
 					Protocol:   corev1.ProtocolTCP,
 					Port:       int32(endpoint.TargetPort),
 					TargetPort: intstr.FromInt(endpoint.TargetPort),
-				}
-				servicePorts = append(servicePorts, servicePort)
+				})
 			}
 		}
 	}
 
+	return service
+}
+
+func getServicesForEndpoints(endpoints map[string]controllerv1alpha1.EndpointList, meta WorkspaceMetadata) []corev1.Service {
```

Comment:
Is this func needed?
Is so, should it be public?

Reply:
This is the original "shape" of the function and as such is used in a number of places in the codebase. l didn't want to introduce more change than required so I just reimplemented this existing method using the new one for consistency but left it here.

I don't think it should be public because it hides too many "knobs" and as such functions more as a utility method for the code in the package than a library method to be exposed.